### ISSUE TITLE
[rllib] Partially evaluating custom model, selecting action, then continuing model evaluation

### ISSUE THREAD

**Issue by internetcoffeephone** [ray-project/ray#8265]
I'm building a custom model with multiple output heads: one actor-critic output (value function, action distribution) and several others. These latter heads depend take the selected action as input.

To list the heads:
- A: Actor-Critic output head
- B: Addtional output head, depends on output of A
- C, D: Additional output heads, depend on output of B

The only thing that comes close to this is the [autoregressive action distribution](https://docs.ray.io/en/latest/rllib-models.html#autoregressive-action-distributions), but I'm not making actions conditional on each other. Instead, B, C and D evaluate rewards.

Is there a way to do this - obtaining the action before the model fully finishes evaluating? Or alternatively, call sub-models B-C-D from postprocess_fn?

*Ray version: 0.8.4*

**Comment by internetcoffeephone** [comment]
@eugenevinitsky Tagging you because you've worked on something similar - though your model only had A and B, where B is the MOA in the [social influence paper](https://github.com/eugenevinitsky/sequential_social_dilemma_games).

Despite the fact that your MOA implementation did not require any additional model evaluations, do you have any ideas on how to tackle the above problem?

**Comment by internetcoffeephone** [comment]
I just realized that I could probably do this in the loss function, where there's access to all relevant parameters. I will try this later and report back

**Comment by internetcoffeephone** [comment]
> I just realized that I could probably do this in the loss function, where there's access to all relevant parameters. I will try this later and report back

This was incorrect.

Instead, I did the relevant calculations one timestep later, and shifted all the values by one timestep in loss calculations. Any state I needed from A on the next timestep was moved to the model state output, which let me access it one timestep later. Now, A and B are evaluated in the same pass, and C/D can be as well (although I haven't done this yet in the commit below).

To any future readers with the same problem, see [this commit](https://github.com/internetcoffeephone/sequential_social_dilemma_games/commit/dd551f7bd0e938b762c0bcff9fca962cf7973543) if you want some pointers.
